packages/expo-updates/ios/EXUpdates/UpdatesStateMachine.swift

```diff
@@ -0,0 +1,306 @@
+}
+
+/**
+ All the possible types of events that can be sent to the machine. Each event
+ will cause the machine to transition to a new state.
+ */
+internal enum UpdatesStateEventType: String {
+  case check
+  case checkCompleteUnavailable
+  case checkCompleteAvailable
+  case checkError
+  case download
+  case downloadComplete
+  case downloadError
+  case restart
+}
+
+// MARK: - Data structures
+
+/**
+ Structure representing an event that can be sent to the machine.
+ Convenience getters are provided to get derived properties that will be
+ used to modify the context when the machine processes an event.
+ */
+internal struct UpdatesStateEvent {
+  var type: UpdatesStateEventType
+  var body: [String: Any] = [:]
+  var manifest: [String: Any]? {
+    guard let manifest = self.body["manifest"] as? [String: Any] else {
+      return nil
+    }
+    return manifest
+  }
+  var error: Error? {
+    guard let message = self.body["message"] as? String else {
+      return nil
+    }
+    return UpdatesStateException(message)
+  }
+  var isRollback: Bool {
+    guard let isRollback = self.body["isRollBackToEmbedded"] as? Bool else {
+      return false
+    }
+    return isRollback
+  }
+  var changedProperties: [String] {
+    return UpdatesStateEvent.updatesStateEventChangedProperties[type] ?? UpdatesStateContext.allProps
+  }
+
+  /**
+   For each event type, an array with the names of the context properties that change
+   when that event is processed.
+   */
+  static let updatesStateEventChangedProperties: [UpdatesStateEventType: [String]] = [
+    .check: ["isChecking"],
+    .checkCompleteAvailable: ["isChecking", "isUpdateAvailable", "checkError", "latestManifest", "isRollback"],
+    .checkCompleteUnavailable: ["isChecking", "isUpdateAvailable", "checkError", "latestManifest", "isRollback"],
+    .checkError: ["isChecking", "checkError"],
+    .download: ["isDownloading"],
+    .downloadComplete: ["isDownloading", "downloadError", "latestManifest", "downloadedManifest", "isUpdatePending", "isUpdateAvailable"],
+    .downloadError: ["isDownloading", "downloadError"],
+    .restart: ["isRestarting"]
+  ]
+}
+
+/**
+ The state machine context, with information that will be readable from JS.
+ */
+internal struct UpdatesStateContext {
+  var isUpdateAvailable: Bool = false
+  var isUpdatePending: Bool = false
+  var isRollback: Bool = false
+  var isChecking: Bool = false
+  var isDownloading: Bool = false
+  var isRestarting: Bool = false
+  var latestManifest: [String: Any]?
+  var downloadedManifest: [String: Any]?
+  var checkError: Error?
+  var downloadError: Error?
+
+  var json: [String: Any] {
+    return [
+      "isUpdateAvailable": self.isUpdateAvailable,
+      "isUpdatePending": self.isUpdatePending,
+      "isRollback": self.isRollback,
+      "isChecking": self.isChecking,
+      "isDownloading": self.isDownloading,
+      "isRestarting": self.isRestarting,
+      "latestManifest": self.latestManifest ?? NSNull(),
```

Comment:
I believe using NSNull in swift usually isn't necessary. Does using nil not work here? If it doesn't, add a comment indicating why we use NSNull

Reply:
Since we are replacing the entire context on each state change, the NSNull is probably safe to remove, agreed.

Reply:
Done